File: tests/fixtures/python-v0.5.0a4/pikamux/paths.py
from __future__ import annotations

import os
from pathlib import Path


def config_home() -> Path:
    override = os.environ.get("PIKA_CONFIG_HOME")
    if override:
        return Path(override).expanduser()
    base = Path(os.environ.get("XDG_CONFIG_HOME", Path.home() / ".config"))
    return base / "pika"


def state_home() -> Path:
    override = os.environ.get("PIKA_STATE_HOME")
    if override:
        return Path(override).expanduser()
    base = Path(os.environ.get("XDG_STATE_HOME", Path.home() / ".local" / "state"))
    return base / "pika"


def config_path() -> Path:
    return config_home() / "config.json"


def database_path() -> Path:
    override = os.environ.get("PIKA_DB_PATH")
    if override:
        return Path(override).expanduser()
    return state_home() / "pika.db"


def codex_home() -> Path:
    return Path(os.environ.get("CODEX_HOME", Path.home() / ".codex")).expanduser()


def claude_home() -> Path:
    return Path(
        os.environ.get("CLAUDE_CONFIG_DIR", Path.home() / ".claude")
    ).expanduser()


def opencode_data_home() -> Path:
    override = os.environ.get("OPENCODE_DATA_HOME")
    if override:
        return Path(override).expanduser()
    base = Path(os.environ.get("XDG_DATA_HOME", Path.home() / ".local" / "share"))
    return base / "opencode"


def opencode_config_home() -> Path:
    override = os.environ.get("OPENCODE_CONFIG_DIR")
    if override:
        return Path(override).expanduser()
    base = Path(os.environ.get("XDG_CONFIG_HOME", Path.home() / ".config"))
    return base / "opencode"

File: src/paths.rs
use anyhow::{Context, Result};
use std::path::PathBuf;

#[derive(Clone, Debug)]
pub struct Paths {
    pub config_dir: PathBuf,
    pub state_dir: PathBuf,
    pub config: PathBuf,
    pub database: PathBuf,
    pub codex_home: PathBuf,
    pub claude_home: PathBuf,
    pub opencode_data_home: PathBuf,
    pub opencode_config_home: PathBuf,
}

impl Paths {
    pub fn discover() -> Result<Self> {
        let home = directories::BaseDirs::new().context("cannot determine home directory")?;
        let config_dir = std::env::var_os("PIKA_CONFIG_HOME")
            .map(PathBuf::from)
            .or_else(|| std::env::var_os("XDG_CONFIG_HOME").map(|p| PathBuf::from(p).join("pika")))
            .unwrap_or_else(|| home.home_dir().join(".config/pika"));
        let state_dir = std::env::var_os("PIKA_STATE_HOME")
            .map(PathBuf::from)
            .or_else(|| std::env::var_os("XDG_STATE_HOME").map(|p| PathBuf::from(p).join("pika")))
            .unwrap_or_else(|| home.home_dir().join(".local/state/pika"));
        let database = std::env::var_os("PIKA_DB_PATH")
            .map(PathBuf::from)
            .unwrap_or_else(|| state_dir.join("pika.db"));
        let codex_home = std::env::var_os("CODEX_HOME")
            .map(PathBuf::from)
            .unwrap_or_else(|| home.home_dir().join(".codex"));
        let claude_home = std::env::var_os("CLAUDE_CONFIG_DIR")
            .map(PathBuf::from)
            .unwrap_or_else(|| home.home_dir().join(".claude"));
        let opencode_data_home = std::env::var_os("OPENCODE_DATA_HOME")
            .map(PathBuf::from)
            .or_else(|| {
                std::env::var_os("XDG_DATA_HOME").map(|p| PathBuf::from(p).join("opencode"))
            })
            .unwrap_or_else(|| home.home_dir().join(".local/share/opencode"));
        let opencode_config_home = std::env::var_os("OPENCODE_CONFIG_DIR")
            .map(PathBuf::from)
            .or_else(|| {
                std::env::var_os("XDG_CONFIG_HOME").map(|p| PathBuf::from(p).join("opencode"))
            })
            .unwrap_or_else(|| home.home_dir().join(".config/opencode"));
        Ok(Self {
            config: config_dir.join("config.json"),
            config_dir,
            state_dir,
            database,
            codex_home,
            claude_home,
            opencode_data_home,
            opencode_config_home,
        })
    }
}
